libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -24,666 +24,849 @@
-    _batt_remaining(0),
-    _batt_volts(0),
-    _batt_amps(0),
-    _sats_data_ready(false),
-    gps_sats(0),
-    _gps_data_ready(false),
-    _pos_gps_ok(false),
-    _course_in_degrees(0),
-    _lat_ns(0),
-    _lon_ew(0),
-    _latdddmm(0),
-    _latmmmm(0),
-    _londddmm(0),
-    _lonmmmm(0),
-    _alt_gps_meters(0),
-    _alt_gps_cm(0),
-    _speed_in_meter(0),
-    _speed_in_centimeter(0),
-    _baro_data_ready(false),
-    _baro_alt_meters(0),
-    _baro_alt_cm(0),
-    _mode_data_ready(false),
-    _mode(0),
-    _fas_call(0),
-    _gps_call(0),
-    _vario_call(0),
-    _various_call(0),
-    _sport_status(0)
+    _ap(),
+    _mav(),
+    _current_height(0.0f),
+    _gps()
     {}
 
-// init - perform require initialisation including detecting which protocol to use
-void AP_Frsky_Telem::init(const AP_SerialManager& serial_manager)
-{
-    // check for FRSky_DPort
-    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FRSky_DPort, 0))) {
-        _protocol = FrSkyDPORT;
-        _initialised_uart = true;   // SerialManager initialises uart for us
-    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FRSky_SPort, 0))) {
-        // check for FRSky_SPort
-        _protocol = FrSkySPORT;
-        _gps_call = 0;
-        _fas_call = 0;
-        _vario_call = 0 ;
-        _various_call = 0 ;
-        _gps_data_ready = false;
-        _battery_data_ready = false;
-        _baro_data_ready = false;
-        _mode_data_ready = false;
-        _sats_data_ready = false;
-        _sport_status = 0;
-        hal.scheduler->register_io_process(FUNCTOR_BIND_MEMBER(&AP_Frsky_Telem::sport_tick, void));
+/*
+ * init - perform required initialisation
+ * for Copter
+ */
+void AP_Frsky_Telem::init(const AP_SerialManager &serial_manager, const char *firmware_str, const char *frame_config_str, const uint8_t mav_type, AP_Float *fs_batt_voltage, AP_Float *fs_batt_mah, uint8_t *control_mode, uint32_t *ap_value, uint32_t *control_sensors_present, uint32_t *control_sensors_enabled, uint32_t *control_sensors_health, int32_t *home_distance, int32_t *home_bearing)
+{
+    // check for protocol configured for a serial port - only the first serial port with one of these protocols will then run (cannot have FrSky on multiple serial ports)
+    if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_D, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_D; // FrSky D protocol (D-receivers)
+    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_SPort; // FrSky SPort protocol (X-receivers)
+    } else if ((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_FrSky_SPort_Passthrough, 0))) {
+        _protocol = AP_SerialManager::SerialProtocol_FrSky_SPort_Passthrough; // FrSky SPort and SPort Passthrough (OpenTX) protocols (X-receivers)
```

Comment:
Initially, our pull request included Rover and Plane, but with all the changes, we've focused on getting Copter done first, then we'll work on adding Plane/Rover.